src/pagecache/iobuf.rs

```diff
@@ -191,10 +191,104 @@ impl IoBuf {
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct StabilityIntervals {
+    fsynced_ranges: Vec<(Lsn, Lsn)>,
+    batches: Vec<(Lsn, Lsn)>,
+    stable_lsn: Lsn,
+}
+
+impl StabilityIntervals {
+    fn new(lsn: Lsn) -> StabilityIntervals {
+        StabilityIntervals {
+            stable_lsn: lsn,
+            fsynced_ranges: vec![],
+            batches: vec![],
+        }
+    }
+
+    pub(crate) fn mark_batch(&mut self, interval: (Lsn, Lsn)) {
+        self.batches.push(interval);
+        // reverse sort
+        self.batches.sort_unstable_by(|a, b| b.cmp(a));
+    }
+
+    fn mark_fsync(&mut self, interval: (Lsn, Lsn)) -> Option<Lsn> {
+        trace!(
+            "pushing interval {:?} into fsynced_ranges {:?}",
+            interval,
+            self.fsynced_ranges
+        );
+        if let Some((low, high)) = self.fsynced_ranges.last_mut() {
+            if *low == interval.1 + 1 {
+                *low = interval.0
+            } else if *high + 1 == interval.0 {
+                *high = interval.1
+            } else {
+                self.fsynced_ranges.push(interval);
+            }
+        } else {
+            self.fsynced_ranges.push(interval);
+        }
+
+        #[cfg(any(test, feature = "event_log", feature = "lock_free_delays"))]
+        assert!(
+            self.fsynced_ranges.len() < 10000,
+            "intervals is getting strangely long... {:?}",
+            self
+        );
+
+        // reverse sort
+        self.fsynced_ranges.sort_unstable_by(|a, b| b.cmp(a));
+
+        while let Some(&(low, high)) = self.fsynced_ranges.last() {
+            assert!(low <= high);
+            let cur_stable = self.stable_lsn;
+            assert!(
+                low > cur_stable,
+                "somehow, we marked offset {} stable while \
+                 interval {}-{} had not yet been applied!",
+                cur_stable,
+                low,
+                high
+            );
+            if cur_stable + 1 == low {
+                debug!("new highest interval: {} - {}", low, high);
+                self.fsynced_ranges.pop().unwrap();
+                self.stable_lsn = high;
+            } else {
+                break;
+            }
+        }
+
+        let mut batch_stable_lsn = None;
+
+        while let Some(&(low, high)) = self.batches.last() {
+            assert!(
+                low < high,
+                "expected batch low mark {} to be below high mark {}",
+                low,
+                high
+            );
+            if high <= self.stable_lsn {
+                if let Some(bsl) = batch_stable_lsn {
+                    assert!(bsl < high);
+                }
+                batch_stable_lsn = Some(high);
+                self.batches.pop().unwrap();
+            } else {
+                break;
+            }
+        }
+
+        if self.batches.is_empty() {
+            Some(self.stable_lsn)
         } else {
-            Err(res)
+            batch_stable_lsn
```

Comment:
It looks like this method will return None if there is one batch being tracked and it is not yet stable. Is that intended, or would it make sense to return the lesser of self.stable_lsn or one less than the first batch's low LSN?

Reply:
I suppose the more general version of that question would be, why is batch_stable_lsn set based on the end of a stabilized batch, rather than just before the beginning of the first non-stable batch? (I may be making some faulty assumptions about what the variables mean)

Reply:
that's a good point, I think it should be bumped up to just before the earliest batch if that's less than or equal to the internal contiguous stable point. I should also improve the comments in that area. 

A goal is to put this into a more formal model of constraints that can be linked down to code, where things like this can be reasoned about more mechanically so stuff like this will hopefully shake out in these correctness critical pathways